AsyncTwoPhaseIndexerTests testStateMachine CI failures

Several failures with the error:

```
java.lang.AssertionError: 
Expected: <6>
     but: was <5>
	at __randomizedtesting.SeedInfo.seed([F2BBC735F05D6D51:9E12BB2E88D0733F]:0)
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.junit.Assert.assertThat(Assert.java:956)
	at org.junit.Assert.assertThat(Assert.java:923)
	at org.elasticsearch.xpack.core.indexing.AsyncTwoPhaseIndexerTests.testStateMachine(AsyncTwoPhaseIndexerTests.java:224)
```

Does _not_ reproduce:

```
./gradlew :x-pack:plugin:core:unitTest \
  -Dtests.seed=F2BBC735F05D6D51 \
  -Dtests.class=org.elasticsearch.xpack.core.indexing.AsyncTwoPhaseIndexerTests \
  -Dtests.method="testStateMachine" \
  -Dtests.security.manager=true \
  -Dtests.locale=en-ZM \
  -Dtests.timezone=Atlantic/Reykjavik \
  -Dcompiler.java=12 \
  -Druntime.java=11
```

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-unix-compatibility/os=centos-6/320/console
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+7.x+multijob-unix-compatibility/os=ubuntu-18.04&&immutable/95/console
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+7.x+multijob-unix-compatibility/os=ubuntu-18.04&&immutable/94/console

--- a/x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/indexing/AsyncTwoPhaseIndexerTests.java
+++ b/x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/indexing/AsyncTwoPhaseIndexerTests.java
@@ -39,7 +39,7 @@ public class AsyncTwoPhaseIndexerTests extends ESTestCase {
 
         private final CountDownLatch latch;
         // test the execution order
-        private int step;
+        private volatile int step;
 
         protected MockIndexer(Executor executor, AtomicReference<IndexerState> initialState, Integer initialPosition,
                               CountDownLatch latch) {
@@ -113,8 +113,8 @@ public class AsyncTwoPhaseIndexerTests extends ESTestCase {
         protected void onFinish(ActionListener<Void> listener) {
             assertThat(step, equalTo(4));
             ++step;
-            isFinished.set(true);
             listener.onResponse(null);
+            isFinished.set(true);
         }
 
         @Override
@@ -206,8 +206,7 @@ public class AsyncTwoPhaseIndexerTests extends ESTestCase {
         }
     }
 
-    @AwaitsFix( bugUrl = "https://github.com/elastic/elasticsearch/issues/40946")
-    public void testStateMachine() throws InterruptedException {
+    public void testStateMachine() throws Exception {
         AtomicReference<IndexerState> state = new AtomicReference<>(IndexerState.STOPPED);
         final ExecutorService executor = Executors.newFixedThreadPool(1);
         isFinished.set(false);

